### DISCUSSION TITLE
arguments passed to `Message::react`

### DISCUSSION THREAD

**Discussion by bunchofcellulose** [serenity-rs/serenity#1854] [Q&A] [answered]
The signature of the function is:
```rs
pub async fn react(
        &self,
        cache_http: impl CacheHttp,
        reaction_type: impl Into<ReactionType>,
    ) -> Result<Reaction>
```
What do you pass into the reaction_type parameter? Please explain with an example

**Comment by arqunis** [comment, marked as answer]
The job of `ReactionType` is to handle two kinds of emoji types:
- Unicode -- these are emojis that exist within the Unicode standard, which as a result, are represented as characters. When the Discord client encounters these characters, they're swapped with the emoji image they represent. To be specific, the characters are replaced with Twitter's https://twemoji.twitter.com/.
- Custom -- these are emojis unique to a guild, represented in a message in the format of `<a?:id:name>` (the `?` means that "a" may be omitted). Custom emojis are just regular static (jpg, png) or animated (gif) images.

`Message::react` uses the `Into` trait to allow users to pass emoji data into the function without having to manually construct `ReactionType`. The kinds of data you may pass depends on `ReactionType`'s `From` implementations, which you may find at https://serenity-rs.github.io/serenity/next/serenity/model/channel/enum.ReactionType.html#impl-From%3CEmoji%3E

As an example, here is how you would react to a message with a Unicode emoji:
```rust
let ctx: &Context = ...;
let msg: &Message = ...;

msg.react(ctx, '😄').await?;
```
And here's with a custom emoji (assuming `1234` is a valid emoji ID and so is `foo` a name, and the emoji itself isn't animated):
```rust
use serenity::model::id::EmojiId;
use serenity::model::misc::EmojiIdentifier;

let ctx: &Context = ...;
let msg: &Message = ...;

msg.react(ctx, EmojiIdentifier { animated: false, id: EmojiId(1234), name: "foo".to_string() }).await?;
```
Of course, this doesn't stop you from using `ReactionType` directly:
```rust
let ctx: &Context = ...;
let msg: &Message = ...;

msg.react(ctx, ReactionType::Unicode("🐱".to_string())).await?;
```
